### ISSUE TITLE
Solved

### ISSUE THREAD

**Issue by user-1701** [nslyubaykin/relax#1]
Hello Nikita,

first of all I would like to acknowledge that this is probably the cleanest ml code I have srsly ever seen. Thank you for that!

So I am using FRWR and Random Shooting and I have a question about the horizon.
I am running on limited resources in real-time (gtx1060+win+cuda, overall I would say RS is quite fast, CEM slow, FRWR a bit slow) and I see that the length of the horizon scales worse in regards to the overall performance. So I experimented to lower the iterations/s and reduced f.i. ensemble size to 3, candidate_sequences = 300, horizon = 3-5. So I thought, it would be nicer, to spread the sampling into the future by 2^x (so horizon[1,2,4,8,16] ), maybe also weighting the later even higher - and discarding the intermediate steps for calculations of candidates. So I could span a bigger time horizon with the same number of steps. (Maybe this approach is already used in the mpc community, but I am fairly new to it, so I don't know.)
I see you have a very interesting scheduler system, but I am confused how to initialize an exponential scheduler that lags the horizon in its tensor dimensions. I made a drawing for this.

![mpc horizon mod](https://github.com/nslyubaykin/relax/assets/12846762/ac13c2c4-fc90-4a2d-bdab-6a157e72a369)

mpc horizon mod

Update: I tried in the data utils get_next_lag_obs

        lag_obs_split = np.split(lag_obs, 
                                 indices_or_sections=1+nlags**2,   # pow  or   =1+int(nlags**1.25)
                                 axis=concat_axis)

I am also thinking to convert all lists to np.arrays and jax jit-ing the buffers (and trying hugginface accelerate), but not sure

Also a short quesstion, would it be possible to replace RS/CEM by [(NN-approximated) MPPI](https://github.com/UM-ARM-Lab/pytorch_mppi))? Because they say so. (A combination of MPPI/FRWR)

Anyways, thank you for your beautiful and very instructive codebase! (I am wondering at what / what instance taught this way of ('pure pythonic'?) coding?)

Best

Lee
Living Computation Foundation Member

**Comment by nslyubaykin** [comment]
Hello! Thank you for your comments. Just saw your question yesterday, had a very busy week. If it is still valid for you I can try to answer it ;-)

**Comment by user-1701** [comment]
Sure, cool! I didn't want to bother and don't feel pressured! 

For my experimentation I didn't have much success with FRWR, MBRL and currently random shooting that is running a lot faster, but that is also not converging well. Compared to an SAC+ICM intrinsic curiousity module, which is also not so fast, but seems to do a little better in finding trajectories. (The ICM is from [here](https://github.com/RLE-Foundation/rllte) and the only acceptable fast one from the intrinsic modules.). I am currently running the SACs on about 2 it/s and PPO on 8-10 it/s.

Regarding the speed, I sometimes wonder why the standard implementations of sac/ppo [this repo](https://github.com/quantumiracle/Popular-RL-Algorithms) ~~are I think faster than anything else i tried (sac 10-20 it/s, ppo 20+ it/s~~, also the ReplayBuffer is just lists. (But maybe I have a bug, as these are not converging anyhow.) I mean, FRWR/RS are also very fast 10+ it/s if the horizon is as low as 3-4, which reduces overall performance significantly I guess.
There are also these evolutionary CEM methods like [GRAC](https://github.com/stanford-iprl-lab/GRAC), but these are horribly slow.

Regarding the MPPI I've seen that the torch library does even have a MPPIPlanner and CEMPlanner, though I assume that the `actions` should be generated by the neural network.

But I mean, the approaches are endless. [Poplin](https://github.com/WilsonWangTHU/POPLIN) which is already 4 years old,  [TDMPC](https://github.com/nicklashansen/tdmpc2), sine infused networks, sine activation functions, various controllers [Snake](https://github.com/zhenshan-bing/RL_Snake). (I built a balancing 2-leg and it is hard to get the cyclical movement extrapolated, though I experimented a lot with reward functions)
([This](https://github.com/google-research/google-research/tree/master/action_angle_networks) looks interesting)
I personally found this quite interesting that claims to have learned the door opener task without a reward function [SkewFit](https://github.com/rail-berkeley/rlkit/blob/master/docs/SkewFit.md)

Me personally, I would just feed a linear function as actions, (as opposed to f.i. Ornstein-Uhlenbeck noise or even Pink Noise) and varying these linear functions, as these are just what my servos should do as trajectory, but I don't know how to make the linear functions being controlled by the network to 'close the loop'. Hm.. maybe by multiplying the mean of a gaussian by the linear function, if I think about it now. Randomly switching in a range of 2 seconds the sign and strength of the slope. This seems to be the most obvious way to define a trajectory (over servo angles, or foot velocities, and maybe generally most trajectories). 

![2023-12-05 16_23_34-Project52 - Unreal Editor](https://github.com/nslyubaykin/relax/assets/12846762/09a1dca7-cb63-44ba-93c4-8962be4429f4)

**Comment by user-1701** [comment]
So I am experimenting with optimizers (natural gradient descent, particle swarm), trying prelu activation and trying the linear actuation with offline rl now. After 4 days the RS might even converge a bit from a gaussian policy, along to sac, but we'll see.

Regarding the Horizon modification, I don't think, that I totally understood the MPC principle and I would assume that it is not possible to simulate further steps in the future, as there is no time distance between prediction steps (which is not necessary if I think about it now). Or if the MPC samples trajectories from the past, it might be possible though. I also think that this idea would have been intensively studied in the field, if it would be that simple.
